---
title:  |
  | **Phacelia hastata**
  | <span style='font-size: 18pt'>*Filtering workflow & exploration of parameters*</span>
output: 
  html_document:
    theme: cosmo
    highlight: tango
---

```{r, echo=FALSE}
library(knitr)
opts_chunk$set(echo = F, message = F, warning = F, fig.width = 10, fig.height = 10, fig.align = 'center')
```

```{r}
library(tidyr)
library(dplyr)
library(stringr)
library(ggplot2)
library(ggrepel)
library(patchwork)
library(MoMAColors)
library(readr)
library(data.table)
library(colorspace)
library(ggnewscale)
library(kableExtra)
```

Initially filtering on 3 things:

+ SNPs only (no indels) `--remove-indels`
+ biallelic sites only `--min-alleles 2` and `--max-alleles 2`
+ 1 SNP per contig (helps account for LD, our reads are typically <100 BP length) `--thin 100`
+ **OPTIONAL (but I typically do)** - filter on call quality. Our call qualities are typically very good so it doesn't get rid of much but might as well cut out the worst stuff `--minQ 30`

```{bash, echo=TRUE, eval=FALSE}
vcftools \
--gzvcf PHHA.vcf.gz \
--remove-indels \
--min-alleles 2 \
--max-alleles 2 \
--thin 100 \
--minQ 30 \
--remove-filtered-all \
--recode \
--recode-INFO-all \
--out PHHA.bithin.q30
```

Produces file **PHHA.bithin.q30.recode.vcf**

Making a summary file to look at individual missingness

```{bash, echo=TRUE, eval=FALSE}
vcftools \
--vcf PHHA.bithin.q30.recode.vcf \
--missing-indv \
--out PHHA.bithin.q30
```

Produces file called **PHHA.bithin.q30.imiss**
  
**256 of 256 individuals**

**669,819 SNPs**

```{r, fig.width=6, fig.height=5}
col37 <- c("#a1def0", "#b32728", "#4ed31b", "#bf209e", "#3f862d", 
           "#b32df9", "#99d683", "#74398b", "#ead624", "#4853fc", 
           "#fe8f06", "#395f97", "#1cf1a3", "#ed8bc7", "#7c440e", 
           "#c6c0fe", "#444a1e", "#efbba2", "#369094", "#fa756b",
           "#88e99a", "#256b33", "#6dc5dd", "#154975", "#628df2",
           "#874fbf", "#e84fe1", "#eec8f1", "#752e4f", "#c7799a",
           "#2a2bf0", "#aebf8a", "#743502", "#bce333", "#e72525",
           "#fea53b", "#ce1365")

ind_miss  <- read_delim("../filtering/PHHA.bithin.q30.imiss", 
                        delim = "\t",
                        col_names = c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), 
                        skip = 1) %>%
    mutate(Pop = str_split_i(ind, '_', 2))

imissFig1 <- ggplot(ind_miss, 
                   aes(x = fmiss, y = Pop, fill = Pop, color = Pop)) +
  geom_jitter(pch = 21, size = 2,
              position = position_jitter(height = 0.2)) +
  scale_fill_manual(values = col37) +
  scale_color_manual(values = darken(col37, 0.3)) +
  scale_x_continuous(limits = c(0.63, 1)) +
  theme_bw() +
  theme(legend.position = 'none')

imissFig1
```

Just looking for some natural breaks here. Wanting to remove individuals that look like obvious outliers without drastically cutting certain populations or lowering within-population sampling. **0.88** looks like a decent threshold. If we wanted to be more aggressive, another option would be **0.85** (would remove another 11 individuals). Probably can't go much lower without losing `DA` and `CA` populations entirely.

Below showing individual totals for each population based on **0.88** threshold.

```{r, fig.width=6, fig.height=5}
pop_sum <- filter(ind_miss, fmiss < 0.88) %>%
  group_by(Pop) %>%
  summarize(n = n())

imissFig2 <- ggplot(ind_miss, 
                   aes(x = fmiss, y = Pop, fill = Pop, color = Pop)) +
  geom_jitter(pch = 21, size = 2,
              position = position_jitter(height = 0.2),
              alpha = ifelse(ind_miss$fmiss > 0.88, 0.25, 1)) +
  geom_vline(xintercept = 0.88, color = 'red',
             linetype = 'dashed', linewidth = 0.5) +
  geom_text(data = pop_sum, aes(y = Pop, label = n),
            x = 0.64,
            size = 4,
            fontface = 'bold') +
  scale_fill_manual(values = col37) +
  scale_color_manual(values = darken(col37, 0.3)) +
  scale_x_continuous(limits = c(0.63, 1)) +
  theme_bw() +
  theme(legend.position = 'none')

imissFig2
```

Individual depth (averaged across all loci). Filtered out individuals >88% missigness as identified above.

**245 of 256 individuals**

**669,819 SNPs**

```{r, fig.width=10, fig.height=4}
var_depth <- read_delim("../filtering/PHHA.bithin.q30.i88.ldepth.mean", delim = "\t",
           col_names = c("chr", "pos", "mean_depth", "var_depth"), skip = 1)

a_freq <- read_delim("../filtering/PHHA.bithin.q30.i88.frq", delim = "\t",
                     col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), skip = 1) %>%
  mutate(fmiss = (max(nchr)-nchr)/max(nchr))

full_depth <- full_join(var_depth, a_freq, by = c("chr", "pos")) %>%
  mutate(depth_adj = mean_depth*(1/(1-fmiss)))

colors <- moma.colors('OKeeffe', type = 'continuous', n = 200)

adpth_all <- ggplot(full_depth, aes(x = fmiss, y = depth_adj)) +
  geom_point(alpha = 0.25, pch = 16, size = 0.3, color = 'black') +
  annotate(geom = 'rect', color = 'red', linewidth = 0.5,
           xmin = -0.001, xmax = 0.5, ymin = 0, ymax = 25, fill = 'transparent') +
  scale_y_continuous(limits = c(0, 100)) +
  theme_bw() +
  theme(legend.position = 'none')

var_adj <- ggplot(full_depth, aes(x = fmiss, y = depth_adj)) +
  geom_point(alpha = 0.05, pch = 16, size = 0.6, color = 'black') +
  geom_density_2d_filled(alpha = 0.6, bins = 200, contour_var = 'count') +
  scale_fill_manual(values = colors) +
  scale_y_continuous(limits = c(0, 25)) +
  scale_x_continuous(limits = c(0, 0.5)) +
  theme_bw() +
  theme(legend.position = 'none')

adpth_all + var_adj
```

```{r, eval=FALSE, echo=TRUE}
snp_set <- filter(full_depth, depth_adj < 12) %>%
  select(chr, pos)

write_delim(snp_set, '../filtering/sites_maxdp12.tsv', delim = '\t', col_names = F)
```

```{r, fig.height=6, fig.width=8}
var_freq <- filter(full_depth, depth_adj < 12)
var_freq$maf <- var_freq %>% dplyr::select(a1, a2) %>% apply(1, function(z) min(z))

maf_set <- matrix(nrow = 51, ncol = 20)
maf_set[,1] <- seq(0, 0.50, 0.01)
amiss_set <- seq(95, 5, -5)
for(i in 1:nrow(maf_set)){
  for(j in 1:length(amiss_set)){
    maf_set[i,(j+1)] <- as.integer(nrow(filter(var_freq, maf > ((i-1)/100) & fmiss < (amiss_set[j]/100))))
  }
}
ind_with <- as.data.frame(maf_set)

fNames <- vector()
for(i in 1:length(amiss_set)){
  fNames[i] = paste('fmiss', as.character(amiss_set[i]), sep = '_')
}
colnames(ind_with)[2:20] <- fNames

comb <- ind_with %>%
  pivot_longer(cols = 2:20, names_to = 'fmiss', values_to = 'snps')

param_compare <- ggplot(comb, aes(x = V1, y = snps, group = fmiss, label = fmiss, color = fmiss)) +
  geom_line(linewidth = 1) +
  scale_x_continuous(breaks = seq(-0.02, 0.5, 0.02)) +
  scale_y_continuous(breaks = seq(0, 100000, 10000)) +
  coord_cartesian(xlim = c(-0.02, 0.11),
                  ylim = c(0, 50000)) +
  geom_label_repel(size = ifelse(comb$V1 == 0, 4, 0),
                   alpha = ifelse(comb$V1 == 0, 1, 0),
                   color = 'black',
                   max.overlaps = 40,
                   label.padding = 0.1, box.padding = 0.65) +
  scale_color_moma_d(palette_name = 'Warhol') +
  labs(x = 'MAF',
       y = 'Total Loci') +
  ggtitle('Loci Retention Rate') +
  theme_light() +
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = 'none')

param_compare
```

```{r}
makePopId <- function(fileIndv){
  PopIDdf = read.table(fileIndv, sep="\t") %>%
    as.data.frame() %>%
    rename(All = V1) %>%
    mutate(Pop = str_split_i(All, '_', 2))
  return(PopIDdf)
}

pca1234FUN <- function(file012, popID, colorSet, grouping){
  df012 = fread(file012, sep = '\t', data.table = F)[,-1] %>%
    as.matrix() %>%
    apply(2, function(d) gsub(-1, NA, d, fixed=TRUE)) %>%
    apply(2, function(d) as.numeric(d))
  colmean = apply(df012, 2, mean, na.rm=TRUE)
  normalize = matrix(nrow = nrow(df012), ncol = ncol(df012))
  af = colmean/2
  for (m in 1:length(af)){
    nr = df012[ ,m]-colmean[m]
    dn = sqrt(af[m]*(1-af[m]))
    normalize[ ,m] = nr/dn
  }
  normalize[is.na(normalize)] = 0
  pca012 = prcomp(normalize, scale. = FALSE, center = FALSE)
  pca012_loading = pca012$x
  pc1var = formatC(((summary(pca012)$importance[2,1])*100), digits = 1, format = 'f')
  pc2var = formatC(((summary(pca012)$importance[2,2])*100), digits = 1, format = 'f')
  pc3var = formatC(((summary(pca012)$importance[2,3])*100), digits = 1, format = 'f')
  pc4var = formatC(((summary(pca012)$importance[2,4])*100), digits = 1, format = 'f')
  pc5var = formatC(((summary(pca012)$importance[2,5])*100), digits = 1, format = 'f')
  pc6var = formatC(((summary(pca012)$importance[2,6])*100), digits = 1, format = 'f')
  pca_df = pca012_loading[,1:6]
  pca_df = cbind(popID, pca_df) 
  pca12_pop = ggplot(data = pca_df, aes(x = PC1, y = PC2, 
                                        fill = .data[[grouping]],
                                        color = .data[[grouping]])) +
    geom_point(pch = 21, size = 3) + 
    xlab(paste("PC",1," (",pc1var,"%)",sep="")) +
    ylab(paste("PC",2," (",pc2var,"%)",sep="")) +
    scale_fill_manual(name = paste(grouping, ':', sep = ''), values = colorSet) +
    scale_color_manual(name = paste(grouping, ':', sep = ''), values = darken(colorSet, 0.3)) +
    theme_bw() + 
    theme(legend.position = 'none',
          axis.text = element_text(size = 13),
          axis.title = element_text(size = 16, color = "black",
                                    face = "bold", vjust = 1),
          panel.border = element_rect(linewidth = 1.5, color = "black"),
          legend.text = element_text(size = 11),
          legend.title = element_text(size = 13, face = "bold"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank())
  pca34_pop = ggplot(data = pca_df, aes(x = PC3, y = PC4,
                                        fill = .data[[grouping]],
                                        color = .data[[grouping]])) +
    geom_point(pch = 21, size = 3) + 
    xlab(paste("PC",3," (",pc3var,"%)",sep="")) +
    ylab(paste("PC",4," (",pc4var,"%)",sep="")) +
    scale_fill_manual(name = paste(grouping, ':', sep = ''), values = colorSet) +
    scale_color_manual(name = paste(grouping, ':', sep = ''), values = darken(colorSet, 0.3)) +
    theme_bw() + 
    theme(legend.position = 'none',
          axis.text = element_text(size = 13),
          axis.title = element_text(size = 16, color = "black",
                                    face = "bold", vjust = 1),
          panel.border = element_rect(linewidth = 1.5, color = "black"),
          legend.text = element_text(size = 11),
          legend.title = element_text(size = 13, face = "bold"),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank())
  finalPatch = pca12_pop + pca34_pop + plot_layout(ncol = 2)
  return(finalPatch)
}

fwStd <- function(file012, popID, xAxis, yAxis, colorSet, grouping){
  df012 = fread(file012, sep = '\t', data.table = F)[,-1] %>%
    as.matrix() %>%
    apply(2, function(d) gsub(-1, NA, d, fixed=TRUE)) %>%
    apply(2, function(d) as.numeric(d))
  colmean = apply(df012, 2, mean, na.rm=TRUE)
  normalize = matrix(nrow = nrow(df012), ncol = ncol(df012))
  af = colmean/2
  for (m in 1:length(af)){
    nr = df012[ ,m]-colmean[m]
    dn = sqrt(af[m]*(1-af[m]))
    normalize[ ,m] = nr/dn
  }
  normalize[is.na(normalize)] = 0
  pca012 = prcomp(normalize, scale. = FALSE, center = FALSE)
  pcXvar = formatC(((summary(pca012)$importance[2,as.integer(str_sub(xAxis, 3, 3))])*100), 
                   digits = 1, format = 'f')
  pcYvar = formatC(((summary(pca012)$importance[2,as.integer(str_sub(yAxis, 3, 3))])*100), 
                   digits = 1, format = 'f')
  pca012_loading = pca012$x
  pca_df = pca012_loading[,1:6]
  pca_df = cbind(popID, pca_df)
  pca_mean <- pca_df %>% 
    group_by(.data[[grouping]]) %>% 
    summarize(PC1_mean=mean(PC1),PC2_mean=mean(PC2),PC3_mean=mean(PC3),
              PC4_mean=mean(PC4),PC5_mean=mean(PC5),PC6_mean=mean(PC6))
  pca_df <- left_join(pca_df,pca_mean)
  pca_fw <- ggplot(data = pca_mean, 
                   aes(x = .data[[paste(xAxis,'mean',sep='_')]],
                       y = .data[[paste(yAxis,'mean',sep='_')]],
                       fill = .data[[grouping]],
                       color = .data[[grouping]], 
                       label = .data[[grouping]])) +
    geom_segment(data=pca_df,
                 aes(x = .data[[xAxis]],y = .data[[yAxis]],
                     xend = .data[[paste(xAxis,'mean',sep='_')]],
                     yend = .data[[paste(yAxis,'mean',sep='_')]]),
                 linewidth = 0.3) +
    geom_label_repel(color = 'black',
                     max.overlaps = 40, label.padding = 0.1, box.padding = 0.65,
                     label.size = 0.1,
                     alpha = 0.5,
                     size = 2,
                     segment.size = 0.1) +
    geom_point(size = 1, 
               stroke = 0.5, pch = 21) + 
    xlab(paste("PC",(str_sub(xAxis, 3, 3))," (",pcXvar,"%)",sep="")) +
    ylab(paste("PC",(str_sub(yAxis, 3, 3))," (",pcYvar,"%)",sep="")) +
    scale_fill_manual(name = paste(grouping, ':', sep = ''), values = colorSet) +
    scale_color_manual(name = paste(grouping, ':', sep = ''), values = darken(colorSet, 0.3)) +
    theme_bw() +
    theme(legend.position = 'none',
          axis.text = element_text(size = 10),
          axis.title.y.left = element_text(size = 16,
                                           face = 'bold',
                                           vjust = 0),
          axis.title.x.bottom = element_text(size = 16,
                                             face = 'bold',
                                             vjust = 1),
          panel.border = element_rect(linewidth = 1.5, color = "black"),
          legend.text = element_text(size = 13),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          plot.margin = margin(t = 5, r = 5, b = 5, l = 5, unit = 'pt'))
  return(pca_fw)
}
```

## Allelic missingness <30% (allele in minimum 172 of 245)

<style>
  hr.thick {
    height: 4px;
    background-color: lightgrey;
    border: 2px solid black;
  }
</style>

<hr>

#### **MAF = 0.01** (minor allele in >= 3 individuals)

#### **15,009 SNPs**

```{r, fig.width=10, fig.height=8}
Apop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf01.012.indv')
Apca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf01.012', Apop, col37, 'Pop')
Afw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf01.012', Apop, 'PC1', 'PC2', col37, 'Pop')
Afw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf01.012', Apop, 'PC3', 'PC4', col37, 'Pop')

A5 <- Apca + Afw + Afw2
A5
```

<hr class="thick">

#### **MAF = 0.02** (minor allele in >= 5 individuals)

#### **8,993 SNPs**

```{r, fig.width=10, fig.height=8}
Gpop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf02.012.indv')
Gpca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf02.012', Gpop, col37, 'Pop')
Gfw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf02.012', Gpop, 'PC1', 'PC2', col37, 'Pop')
Gfw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf02.012', Gpop, 'PC3', 'PC4', col37, 'Pop')

G5 <- Gpca + Gfw + Gfw2
G5
```

<hr class="thick">

#### **MAF = 0.03** (minor allele in >= 8 individuals)

#### **6,811 SNPs**

```{r, fig.width=10, fig.height=8}
Bpop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf03.012.indv')
Bpca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf03.012', Bpop, col37, 'Pop')
Bfw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf03.012', Bpop, 'PC1', 'PC2', col37, 'Pop')
Bfw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf03.012', Bpop, 'PC3', 'PC4', col37, 'Pop')

B5 <- Bpca + Bfw + Bfw2
B5
```

<hr class="thick">

#### **MAF = 0.04** (minor allele in >= 8 individuals)

#### **5,617 SNPs**

```{r, fig.width=10, fig.height=8}
Hpop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf04.012.indv')
Hpca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf04.012', Hpop, col37, 'Pop')
Hfw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf04.012', Hpop, 'PC1', 'PC2', col37, 'Pop')
Hfw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf04.012', Hpop, 'PC3', 'PC4', col37, 'Pop')

H5 <- Hpca + Hfw + Hfw2
H5
```

<hr class="thick">

#### **MAF = 0.05** (minor allele in >= 13 individuals)

#### **4,843 SNPs**

```{r, fig.width=10, fig.height=8}
Cpop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf05.012.indv')
Cpca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf05.012', Cpop, col37, 'Pop')
Cfw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf05.012', Cpop, 'PC1', 'PC2', col37, 'Pop')
Cfw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf05.012', Cpop, 'PC3', 'PC4', col37, 'Pop')

C5 <- Cpca + Cfw + Cfw2
C5
```

<hr class="thick">

#### **MAF = 0.06** (minor allele in >= 13 individuals)

#### **4,276 SNPs**

```{r, fig.width=10, fig.height=8}
Ipop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf06.012.indv')
Ipca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf06.012', Ipop, col37, 'Pop')
Ifw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf06.012', Ipop, 'PC1', 'PC2', col37, 'Pop')
Ifw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a70.maf06.012', Ipop, 'PC3', 'PC4', col37, 'Pop')

I5 <- Ipca + Ifw + Ifw2
I5
```

```{r, fig.width=5, fig.height=5}
snp_var_ratios <- tibble(maf = seq(0.01, 0.06, 0.01),
                         snps = c(15009, 8993, 6811, 5617, 4843, 4276),
                         pc1 = c(0.027, 0.037, 0.043, 0.048, 0.052, 0.054),
                         pc2 = c(0.02, 0.025, 0.028, 0.03, 0.031, 0.032),
                         pc3 = c(0.018, 0.02, 0.022, 0.023, 0.023, 0.023),
                         pc4 = c(0.015, 0.018, 0.019, 0.02, 0.021, 0.021)) %>%
  mutate('pc1+2' = pc1 + pc2,
         'pc1+2+3' = pc1 + pc2 + pc3,
         'pc1+2+3+4' = pc1 + pc2 + pc3 + pc4) %>%
  pivot_longer(cols = c('pc1', 'pc1+2', 'pc1+2+3', 'pc1+2+3+4'), names_to = 'axes',
               values_to = 'var_exp')

second <- distinct(snp_var_ratios, maf, snps)

coeff <- 110000

by1 <- ggplot() +
  geom_line(data = snp_var_ratios, aes(x = maf, y = var_exp, group = axes, color = axes)) +
  geom_line(data = second, aes(x = maf, y = snps / coeff)) +
  geom_point(pch = 16) +
    scale_y_continuous(limits = c(0,0.15),
    name = "Total Variance Explained",
    sec.axis = sec_axis(~.*coeff, name="Total SNPs")
  ) + 
  theme_bw() +
  theme(legend.position = 'bottom')

by1
```


## Allelic missingness <20% (allele in minimum 196 of 245)

<style>
  hr.thick {
    height: 4px;
    background-color: lightgrey;
    border: 2px solid black;
  }
</style>

<hr>

#### **MAF = 0.01** (minor allele in >= 3 individuals)

#### **8,488 SNPs**

```{r, fig.width=10, fig.height=8}
Dpop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf01.012.indv')
Dpca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf01.012', Dpop, col37, 'Pop')
Dfw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf01.012', Dpop, 'PC1', 'PC2', col37, 'Pop')
Dfw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf01.012', Dpop, 'PC3', 'PC4', col37, 'Pop')

D5 <- Dpca + Dfw + Dfw2
D5
```

<hr class="thick">

#### **MAF = 0.03** (minor allele in >= 8 individuals)

#### **4,030 SNPs**

```{r, fig.width=10, fig.height=8}
Epop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf03.012.indv')
Epca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf03.012', Epop, col37, 'Pop')
Efw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf03.012', Epop, 'PC1', 'PC2', col37, 'Pop')
Efw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf03.012', Epop, 'PC3', 'PC4', col37, 'Pop')

E5 <- Epca + Efw + Efw2
E5
```

<hr class="thick">

#### **MAF = 0.05** (minor allele in >= 13 individuals)

#### **2,900 SNPs**

```{r, fig.width=10, fig.height=8}
Fpop <- makePopId('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf05.012.indv')
Fpca <- pca1234FUN('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf05.012', Fpop, col37, 'Pop')
Ffw <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf05.012', Fpop, 'PC1', 'PC2', col37, 'Pop')
Ffw2 <- fwStd('../filtering/PHHA.bithin.q30.i88.maxdp12.a80.maf05.012', Fpop, 'PC3', 'PC4', col37, 'Pop')

F5 <- Fpca + Ffw + Ffw2
F5
```

